SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp

```diff
@@ -0,0 +1,550 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "LocalDeploymentManager.h"
+
+#include "AssetRegistryModule.h"
+#include "Async/Async.h"
+#include "DirectoryWatcherModule.h"
+#include "FileCache.h"
+#include "GeneralProjectSettings.h"
+#include "HAL/FileManager.h"
+#include "HAL/PlatformFilemanager.h"
+#include "Interop/Connection/EditorWorkerController.h"
+#include "Misc/FileHelper.h"
+#include "Serialization/JsonReader.h"
+#include "Serialization/JsonSerializer.h"
+#include "Serialization/JsonWriter.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialDeploymentManager);
+
+FLocalDeploymentManager::FLocalDeploymentManager()
+{
+	bLocalDeploymentRunning = false;
+	bSpatialServiceRunning = false;
+
+	bStartingDeployment = false;
+	bStoppingDeployment = false;
+
+	bStartingSpatialService = false;
+	bStoppingSpatialService = false;
+
+	// For checking whether we can stop or start. Set in the past so the first RefreshServiceStatus does not wait.
+	LastSpatialServiceCheck = FDateTime::Now() - FTimespan::FromSeconds(RefreshFrequency);
+
+	// Get the project name from the spatialos.json.
+	ProjectName = GetProjectName();
+
+	// Ensure the worker.jsons are up to date.
+	WorkerBuildConfigAsync();
+
+	// Watch the worker config directory for changes.
+	StartUpWorkerConfigDirectoryWatcher();
+}
+
+FString FLocalDeploymentManager::GetSpotExe()
+{
+	FString PluginDir = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectPluginsDir(), TEXT("UnrealGDK")));
+
+	if (!FPaths::DirectoryExists(PluginDir))
+	{
+		// If the Project Plugin doesn't exist then use the Engine Plugin.
+		PluginDir = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::EnginePluginsDir(), TEXT("UnrealGDK")));
+		ensure(FPaths::DirectoryExists(PluginDir));
+	}
+
+	return  FPaths::ConvertRelativePathToFull(FPaths::Combine(PluginDir, TEXT("SpatialGDK/Binaries/ThirdParty/Improbable/Programs/spot.exe")));
+}
+
+FString FLocalDeploymentManager::GetSpatialOSDirectory()
+{
+	return FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectDir(), TEXT("/../spatial/")));
```

Comment:
:(

Reply:
Moved to static function in GDKServices. Killed old function in SpatialGDKEditorToolbar.